controlpanel/api/tools.py

```diff
@@ -0,0 +1,237 @@
+
+
+class ToolRepository(UserDict):
+
+    def __getitem__(self, key):
+        if key not in self.tools:
+            raise UnsupportedToolException(key)
+
+        if key not in self.tools:
+            raise UnsupportedToolException(key)
+
+        return super().__getitem__(key)
+
+    def __setitem__(self, key, value):
+        if key not in SUPPORTED_TOOL_NAMES:
+            raise UnsupportedToolException(key)
+
+        if key in self.tools:
+            raise DuplicateTool(key)
+
+        super().__setitem__(key, value)
+
+
+class ToolMeta(type):
+    """Adds Tool subclasses to repository"""
+
+    # repository is shared across all Tool subclasses
+    _repository = ToolRepository()
+
+    def __init__(cls, name, bases, clsdict):
+        super().__init__(name, bases, clsdict)
+        if 'name' not in clsdict:
+            raise AttributeError(f"{name}.name")
+        if name != "Tool":
+            cls._repository[clsdict['name']] = cls
+
+
+class Tool(metaclass=ToolMeta):
+    name = None
+
+    @classmethod
+    def create(cls, name):
+        return cls.repository[name]()
+
+    @cached_property
+    def helm(self):
+        return Helm()
+
+    @property
+    def chart_name(self):
+        return f"mojanalytics/{self.name}"
+
+    def release_name(self, username):
+        return f"{username}-{self.name}"
+
+    def deploy_params(self, user):
+        conf = settings.TOOLS[self.name]
+        return {
+            "username": user.username.lower(),
+            "aws.iamRole": user.iam_role_name,
+            "toolsDomain": settings.TOOLS_DOMAIN,
+            "authProxy.cookieSecret": secrets.token_hex(32),
+            f"{self.name}.secureCookieKey": secrets.token_hex(32),
+            "authProxy.auth0.domain": conf["domain"],
+            "authProxy.auth0.clientId": conf["client_id"],
+            "authProxy.auth0.clientSecret": conf["client_secret"],
+        }
+
+    def deploy_for(self, user):
+        """
+        Deploy the given tool in the user namespace.
+        >>> class RStudio(BaseTool):
+        ...     name = 'rstudio'
+        >>> rstudio = RStudio()
+        >>> rstudio.deploy_for(alice)
+        """
+
+        username = user.username.lower()
+        params = self.deploy_params(user)
+        if params:
+            params = [
+                "--set",
+                ",".join(f"{key}={val}" for key, val in params.items()),
+            ]
+        return self.helm.upgrade_release(
+            self.release_name(username),
+            self.chart_name,
+            f"--namespace={user.k8s_namespace}",
+            *params,
+        )
+
+    def get_user_deployment(self, user):
+        deployments = client.AppsV1Api().list_namespaced_deployment(
+            namespace=user.k8s_namespace,
+            label_selector=f"app={self.name}"
+        )
+
+        if len(deployments.items) < 1:
+            raise ToolNotDeployed()
+
+        return ToolDeployment(deployments.items[0], user)
```

Comment:
#### 1e. Refactor Tools code
> Lines 123 - 132

The main change to the base `Tool` class is to add a method to get the Kubernetes Deployment of the tool for a user.

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-aad7b41260663fb400ff942a02fd3d14R158)